hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/DatanodeUtil.java

```diff
@@ -127,6 +129,31 @@ public static File idToBlockDir(File root, long blockId) {
     return new File(root, path);
   }
 
+  /**
+   * Take an example. We hava a block with blockid mapping to:
+   * "/data1/hadoop/hdfs/datanode/current/BP-xxxx/current/finalized/subdir0/subdir0"
+   * We return "subdir0/subdir0"
+   * @param blockId blockId
+   * @return The two-level subdir name
+   */
+  public static String idToBlockDirSuffixName(long blockId) {
+    int d1 = (int) ((blockId >> 16) & 0x1F);
+    int d2 = (int) ((blockId >> 8) & 0x1F);
+    return DataStorage.BLOCK_SUBDIR_PREFIX + d1 + SEP +
+        DataStorage.BLOCK_SUBDIR_PREFIX + d2;
+  }
+
+  public static List<String> getAllSubDirNameForDataSetLock() {
```

Comment:
I didn't get what this method want to do, and why there are double layer loop (include the invoke enter) and user 0x1F as end condition.

Reply:
This method is used to generate all subdir lock name.  We use 0x1F as end condition because the layout of DataNode.
Each blockpool dir under one volume has two layer subdir, like  subdir1/subdir31.

Reply:
When DataNode starts, we will initial all dataset subdir lock in advance. Every volume has its subdir dataset locks.
In current layout of datanode's  storage directory， we have 32 multipy 32 subdirs in each volume.

Reply:
I don't think this is good idea to follow the current directory layout to construct lock based blockid. since:
a. We have to keep same logic between the directory layout and block/subdir lock, right? However it implements at different place without any constraint, which will involve some cost for the future improvement when change the directory layout. 
b. I think we could decouple block/subdir lock from directory layout, define another rule for block/subdir lock with blockid only(such as `id mod 1000`, split one volume lock to 1000 locks is enough to improve the performance), which I think it make sense to block/subdir.
What do you think about? Thanks.